### ISSUE TITLE
opennmt-py agent

### ISSUE THREAD

**Issue by jsedoc** [facebookresearch/ParlAI#605]
I see that you have some branches with ibm_seq2seq and tf_seq2seq integration. Do you have plans for OpenNMT-py? So far my main issue with this has been the use of torchtext.

Also, for the ibm_seq2seq it is an agent whereas the tf_seq2seq you are using the project folder. Ryo made a tensorflow agent https://github.com/ryonakamura/parlai_agents/blob/master/tensorflow_rnn/tensorflow_rnn.py

**Comment by alexholdenmiller** [comment]
We'd be happy to take a PR for OpenNMT-py! What is the issue with torchtext?

For tf_seq2seq we were trying to get Google's seq2seq code launched, but I had some difficulty integrating it as an agent. Thanks for that reference!

**Comment by jsedoc** [comment]
@alexholdenmiller I'm going to submit a PR, but there's an issue. I needed to make a few changes to your data iterator. Would this be ok? Or do you want me to move it out like you did with TF?

**Comment by alexholdenmiller** [comment]
It depends on the changes :) you'd have to tell me more about it

**Comment by jaseweston** [comment]
ping?

**Comment by jsedoc** [comment]
I've been able to hack up something, but I will try to clean it up for a PR.

**Comment by Henry-E** [comment]
This would be great to see

**Comment by jsedoc** [comment]
So, the main issue is that I need to make a torchtext teacher similar to your pytorch teacher because this is needed for [https://github.com/OpenNMT/OpenNMT-py/blob/master/train.py#L115](url). 

For now, I've hacked around this by making a fake hard-coded fields, and using the ParlAi native teachers. [This may be the easiest, but it is more fragile].

Also, there is significant overlap between torchtext and the data task agents. It might be worth trying to figure out an interface between them. But I'm not knowledgeable enough on ParlAi to take such a large undertaking.

I'll start to clean up for a PR after CoNLL deadline. 

Please drop me a line if you've got a better idea.

**Comment by jsedoc** [comment]
Ok, so June 5th is my most reasonable delivery date. If anyone wants to push faster please feel free. 

My fork is here https://github.com/jsedoc/ParlAI

**Comment by alexholdenmiller** [comment]
Thanks @jsedoc! You're working on a torchtext teacher just to reproduce training on a specific task in the torchtext repo? Which task?

**Comment by Henry-E** [comment]
I'm going to have a look through the repo but I always figured the biggest issue with combining OpenNMT-py and ParlAI was having to reimplement the whole preprocessing step that OpenNMT-py requires. How did you get around that?

**Comment by Henry-E** [comment]
Thinking out loud about a wrapper for OpenNMT-py

Goal is to keep OpenNMT-py's code in a separate repo and minimize the number of modifications needed in OpenNMT-py's repo. Will use subprocess to call OpenNMT-py's command line functions.

- Create a data loading agent that can take any number of observation dicts and convert them to source and target files
- Make source and target files for the whole dataset. Run these files through OpenNMT-py/preprocess.py using subprocess. From the preprocess module output save the vocab dicts
-  For each batch run the source and target files through preprocess, reuse the whole dataset vocab. Call OpenNMT-py/train.py. After the first batch is trained use the most recent checkpoint file to reload the model and continue training

Things to work out
- OpenNMT-py needs a validation set as part of its training but this can probably be a one line file since we don't need to validate every batch
- With regards to doing validation after each epoch or N batches using OpenNMT-py/translate.py. I don't know if the translate module gives perplexity out of the box in the same way it does for the validation set during training
- Training is going to be slow with so much data saving and loading but that's to be expected. Be good to know how much slower it is. But worth it to be able to use copy attention and some other features

**Comment by jaseweston** [comment]
this seems stale so closing for now?
